lib/logstash/codecs/identity_map_codec.rb

```diff
@@ -0,0 +1,255 @@
+# encoding: utf-8
+require "logstash/namespace"
+# require "concurrent"
+require "thread_safe"
+
+# This class is a Codec duck type
+# Using Composition, it maps from a stream identity to
+# a cloned codec instance via the same API as a Codec
+# it implements the codec public API
+
+module LogStash module Codecs class IdentityMapCodec
+  # subclass of Exception, LS has more than limit (20000) active streams
+  class IdentityMapUpperLimitException < Exception; end
```

Comment:
+1 on that reasoning, maybe its time to refactor our base error class too for the next major version WDYT?